### PR TITLE
Use Jdk7DirectoryWalker as default, since we don't support JDK <8 anymore

### PR DESCRIPTION
**Pull request by paplorinc**
Fixes: https://github.com/gradle/gradle/issues/9075 and https://github.com/gradle/gradle/issues/9084

Since we don't support JDK 7 anymore, the alternative walker is not needed.
Removing it opens up the possibility for some minor cleanup and the related tests should also be adjusted.
Also renamed `Jdk7DirectoryWalker` to `DefaultDirectoryWalker` (and changed the package), since it's not related to JDK 7 anymore.

Please review [commit-by-commit](https://github.com/gradle/gradle/pull/9743/commits) to have context for the different changes.

### CODE REVIEW

File: subprojects/files/src/main/java/org/gradle/api/internal/file/collections/DefaultDirectoryWalkerFactory.java [hunk 1]
```diff
@@ -50,20 +47,6 @@ private void reset() {
     }
 
     private DirectoryWalker createInstance() {
-        if (javaVersion.isJava8Compatible() || (javaVersion.isJava7Compatible() && defaultEncodingContainsPlatformEncoding())) {
-            return new Jdk7DirectoryWalker(fileSystem);
-        } else {
-            return new DefaultDirectoryWalker(fileSystem);
```
Comment (paplorinc) [inline subprojects/files/src/main/java/org/gradle/api/internal/file/collections/DefaultDirectoryWalkerFactory.java]: this path should be unused after Java 8 -> deleted

File: subprojects/files/src/main/java/org/gradle/api/internal/file/collections/DirectoryFileTree.java [hunk 2]
```diff
@@ -156,7 +154,7 @@ private void walkDir(File file, RelativePath path, FileVisitor visitor, Spec<Fil
         if (visitor instanceof ReproducibleFileVisitor && ((ReproducibleFileVisitor) visitor).isReproducibleFileOrder()) {
             directoryWalker = REPRODUCIBLE_DIRECTORY_WALKER;
         } else {
-            directoryWalker = directoryWalkerFactory.create();
+            directoryWalker = DEFAULT_DIRECTORY_WALKER;
```
Comment (paplorinc) [inline subprojects/files/src/main/java/org/gradle/api/internal/file/collections/DirectoryFileTree.java]: we don't need a factory for a single instance anymore

File: subprojects/files/src/main/java/org/gradle/api/internal/file/collections/DefaultDirectoryWalker.java [hunk 3]
```diff
@@ -40,10 +39,10 @@
 import java.util.LinkedList;
 import java.util.concurrent.atomic.AtomicBoolean;
 
-public class Jdk7DirectoryWalker implements DirectoryWalker {
```
Comment (paplorinc) [inline subprojects/files/src/main/java/org/gradle/api/internal/file/collections/DefaultDirectoryWalker.java]: not related to JDK 7 anymore

File: subprojects/files/src/main/java/org/gradle/api/internal/file/collections/DefaultDirectoryWalker.java [hunk 4]
```diff
@@ -51,90 +51,111 @@ static boolean isAllowed(FileTreeElement element, Spec<? super FileTreeElement>
...
-                        if (directoryDetailsHolder.size() > 1 && !postfix) {
-                            visitor.visitDir(details);
-                        }
-                        return checkStopFlag();
-                    } else {
-                        return FileVisitResult.SKIP_SUBTREE;
-                    }
-                }
+            PathVisitor pathVisitor = new PathVisitor(directoryDetailsHolder, spec, postfix, visitor, stopFlag, rootPath);
+            Files.walkFileTree(rootDir.toPath(), EnumSet.of(FileVisitOption.FOLLOW_LINKS), Integer.MAX_VALUE, pathVisitor);
```
Comment (paplorinc) [inline subprojects/files/src/main/java/org/gradle/api/internal/file/collections/DefaultDirectoryWalker.java]: simplify structure (reduce depth) by extracting the anonymous inner class.

File: subprojects/files/src/main/java/org/gradle/api/internal/file/collections/DefaultDirectoryWalker.java [hunk 5]
```diff
@@ -52,7 +52,7 @@ static boolean isAllowed(FileTreeElement element, Spec<? super FileTreeElement>
 
     @Override
     public void walkDir(File rootDir, RelativePath rootPath, FileVisitor visitor, Spec<? super FileTreeElement> spec, AtomicBoolean stopFlag, boolean postfix) {
-        Deque<FileVisitDetails> directoryDetailsHolder = new LinkedList<FileVisitDetails>();
+        Deque<FileVisitDetails> directoryDetailsHolder = new ArrayDeque<>();
```
Comment (paplorinc) [inline subprojects/files/src/main/java/org/gradle/api/internal/file/collections/DefaultDirectoryWalker.java]: `ArrayDeque` should scale better than a dispersed linked list.

File: subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryWalkerTest.groovy [hunk 6]
```diff
@@ -29,7 +29,10 @@ import java.util.concurrent.atomic.AtomicInteger
 class DirectoryWalkerTest extends AbstractDirectoryWalkerTest<DirectoryWalker> {
     @Override
     protected List<DirectoryWalker> getWalkers() {
-        return [new DefaultDirectoryWalker(), new ReproducibleDirectoryWalker()]
+        return [
+            new DefaultDirectoryWalker(TestFiles.fileSystem()),
```
Comment (paplorinc) [inline subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryWalkerTest.groovy]: was calling a non-existent default constructor, without initializing the file system parameter

File: subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DefaultUnauthorizedDirectoryWalkerTest.groovy [hunk 7]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2017 the original author or authors.
+ * Copyright 2019 the original author or authors.
```
[the remark this answers is not in the document]
Comment (paplorinc) [reply subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DefaultUnauthorizedDirectoryWalkerTest.groovy]: * this was done automatically by Idea
* it's what I saw in other commits as well (including yours)
* could we delete or change all of these in a single commit and not talk about it anymore? :)

File: subprojects/files/src/testFixtures/groovy/org/gradle/api/internal/file/collections/AbstractDirectoryWalkerTest.groovy [hunk 8]
```diff
@@ -214,7 +215,7 @@ abstract class AbstractDirectoryWalkerTest<T> extends Specification {
         file2 << '12345'
         def file3 = rootDir.createFile("a/b/3.txt")
         file3 << '12345'
-        def walkerInstance = new DefaultDirectoryWalker()
+        def walkerInstance = new DefaultDirectoryWalker(FileSystems.getDefault())
```
[the remark this answers is not in the document]
Comment (paplorinc) [reply subprojects/files/src/testFixtures/groovy/org/gradle/api/internal/file/collections/AbstractDirectoryWalkerTest.groovy]: It's not visible

File: subprojects/files/src/main/java/org/gradle/api/internal/file/collections/DefaultDirectoryWalker.java [hunk 9]
```diff
@@ -51,90 +51,111 @@ static boolean isAllowed(FileTreeElement element, Spec<? super FileTreeElement>
...
+            Files.walkFileTree(rootDir.toPath(), EnumSet.of(FileVisitOption.FOLLOW_LINKS), Integer.MAX_VALUE, pathVisitor);
+        } catch (IOException e) {
+            throw new GradleException(String.format("Could not list contents of directory '%s'.", rootDir), e);
+        }
+    }
 
-                private FileVisitResult checkStopFlag() {
-                    return stopFlag.get() ? FileVisitResult.TERMINATE : FileVisitResult.CONTINUE;
-                }
+    private class PathVisitor implements java.nio.file.FileVisitor<Path> {
```
[the remark this answers is not in the document]
Comment (paplorinc) [reply subprojects/files/src/main/java/org/gradle/api/internal/file/collections/DefaultDirectoryWalker.java]: Added `fileSystem` as a parameter to be able to make it static.

File: subprojects/files/src/main/java/org/gradle/api/internal/file/collections/DefaultDirectoryWalker.java [hunk 10]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2015 the original author or authors.
+ * Copyright 2019 the original author or authors.
```
[the remark this answers is not in the document]
Comment (paplorinc) [reply subprojects/files/src/main/java/org/gradle/api/internal/file/collections/DefaultDirectoryWalker.java]: Done

File: subprojects/files/src/main/java/org/gradle/api/internal/file/collections/DirectoryFileTree.java [hunk 11]
```diff
@@ -49,23 +49,21 @@
...
     private final FileSystem fileSystem;
-    private final Factory<DirectoryWalker> directoryWalkerFactory;
 
     public DirectoryFileTree(File dir, PatternSet patternSet, FileSystem fileSystem) {
-        this(dir, patternSet, DEFAULT_DIRECTORY_WALKER_FACTORY, fileSystem, false);
+        this(dir, patternSet, fileSystem, false);
     }
 
-    DirectoryFileTree(File dir, PatternSet patternSet, Factory<DirectoryWalker> directoryWalkerFactory, FileSystem fileSystem, boolean postfix) {
+    DirectoryFileTree(File dir, PatternSet patternSet, FileSystem fileSystem, boolean postfix) {
```
[the remark this answers is not in the document]
Comment (paplorinc) [reply subprojects/files/src/main/java/org/gradle/api/internal/file/collections/DirectoryFileTree.java]: Done

File: subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DefaultDirectoryWalkerTest.groovy [hunk 12]
```diff
@@ -1,265 +0,0 @@
...
-
-        when:
-        visitor.setExpectedVisitations([[rootFile1, rootFile2], [dir1], [dirFile1, dirFile2]])
-        fileTree.visit(visitor)
-
-        then:
-        visitor.assertExpectations()
-    }
-
-    def walkDepthFirst() {
```
Comment (wolfs) [inline subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DefaultDirectoryWalkerTest.groovy]: I think we should keep these tests and move them to `DefaultDirectoryWalkerTest` maybe?

For example now we don't have any tests for `postfix` traversal any more. Are the other tests covered somewhere else?

Moreover, we need the same tests for `ReproducibleDirectoryWalker`, right?

File: subprojects/files/src/main/java/org/gradle/api/internal/file/collections/ReproducibleDirectoryWalker.java [hunk 13]
```diff
@@ -17,22 +17,71 @@
...
+import org.gradle.api.specs.Spec;
 import org.gradle.internal.nativeintegration.filesystem.FileSystem;
 
 import java.io.File;
+import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class ReproducibleDirectoryWalker implements DirectoryWalker {
```
Comment (wolfs) [inline subprojects/files/src/main/java/org/gradle/api/internal/file/collections/ReproducibleDirectoryWalker.java]: ⭕️ Can we maybe use the Jdk7 walking for the ReproducibleDirectoryWalker as well? I think it would be great to not have two different ways of walking the file system. I think it should be enough to sort the files in one level in `postVisitDirectory` and then notify the listener?

File: subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryWalkerTest.groovy [hunk 14]
```diff
@@ -16,78 +16,22 @@
...
-        where:
-        fileEncoding | expectedClassName
-        "UTF-8"      | "Jdk7DirectoryWalker"
-        "UTF-16be" | "Jdk7DirectoryWalker"
-        "UTF-16le" | "Jdk7DirectoryWalker"
-        "UTF-16"   | "Jdk7DirectoryWalker"
-        "ISO-8859-1" | "DefaultDirectoryWalker"
-    }
-
-    def "both DirectoryWalker implementations return same set of files and attributes"() {
```
Comment (wolfs) [inline subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryWalkerTest.groovy]: We need to keep this test now between `ReproducibleDirectoryWalker` and `DefaultDirectoryWalker`, right?

File: subprojects/files/src/main/java/org/gradle/api/internal/file/collections/ReproducibleDirectoryWalker.java (hunk 13 above)
Comment (paplorinc) [reply subprojects/files/src/main/java/org/gradle/api/internal/file/collections/ReproducibleDirectoryWalker.java]: Sure, does the `⭕️` mean you're okay with doing it in a different PR?

File: subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DefaultDirectoryWalkerTest.groovy (hunk 12 above)
Comment (paplorinc) [reply subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DefaultDirectoryWalkerTest.groovy]: Reverted and fixed, thanks for letting me know they're still relevant.

File: subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryWalkerTest.groovy (hunk 14 above)
Comment (paplorinc) [reply subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryWalkerTest.groovy]: reverted and changed, thanks for the insight.

File: subprojects/files/src/main/java/org/gradle/api/internal/file/collections/ReproducibleDirectoryWalker.java (hunk 13 above)
Comment (wolfs) [reply subprojects/files/src/main/java/org/gradle/api/internal/file/collections/ReproducibleDirectoryWalker.java]: Exactly, the ⭕️ means other PR.

File: subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+package org.gradle.api.internal.file.collections
+
+import org.gradle.api.file.FileVisitDetails
+import org.gradle.api.file.FileVisitor
+import org.gradle.api.internal.file.TestFiles
+import org.gradle.api.specs.Spec
+import org.gradle.api.tasks.util.PatternSet
+import org.gradle.test.fixtures.AbstractProjectBuilderSpec
+
+class DirectoryFileTreeTest extends AbstractProjectBuilderSpec {
```
Comment (wolfs) [inline subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy]: ❌ Can we have this test for `ReproducibleDirectoryWalker` as well?

File: subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy [hunk 16]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2010 the original author or authors.
+ * Copyright 2019 the original author or authors.
```
Comment (wolfs) [inline subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy]: Nit: Copyright year

File: subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy [hunk 17]
```diff
@@ -89,6 +80,7 @@ class DefaultDirectoryWalkerTest extends AbstractProjectBuilderSpec {
 
         Test that the files are really walked breadth first
      */
+
```
Comment (wolfs) [inline subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy]: Why did you add this new line?

File: subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy (hunk 16 above)
Comment (paplorinc) [reply subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy]: Please see: https://github.com/gradle/gradle/pull/9743#discussion_r296441856

File: subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy (hunk 17 above)
Comment (paplorinc) [reply subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy]: Probably automatic formatting during revert

File: subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy (hunk 15 above)
Comment (paplorinc) [reply subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy]: Done

Comment (paplorinc) [reply subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy]: Screen Shot 2019-06-25 at 12 08 14
Screen Shot 2019-06-25 at 12 15 46

File: subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy (hunk 17 above)
Comment (paplorinc) [reply subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy]: Done

File: subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy [hunk 18]
```diff
@@ -259,7 +283,12 @@ class DefaultDirectoryWalkerTest extends AbstractProjectBuilderSpec {
 
         boolean assertExpectations() {
```
Comment (wolfs) [inline subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy]: I think the return type of this method should be `void`, so we don't need to return `true` from it.

Comment (paplorinc) [reply subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy]: yeah, could have been a Spock bug back then, but now voids are okay as well. Done

File: subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy [hunk 19]
```diff
@@ -142,10 +155,15 @@ class DirectoryFileTreeTest extends AbstractProjectBuilderSpec {
 
         then:
         visitor.assertExpectations()
+
+        where:
+        visitor << visitors
     }
 
-    def visitorCanStopVisit() {
+    def "visitor can stop visit - isReproducible: true"() {
```
Comment (paplorinc) [inline subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy]: this needs a given order, so only reproducible makes sense here

Comment (wolfs) [reply subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy]: Can you add the comment in the source code as well?

File: subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy [hunk 20]
```diff
@@ -234,23 +263,26 @@ class DirectoryFileTreeTest extends AbstractProjectBuilderSpec {
...
         private void handleDetails(FileVisitDetails details) {
             def file = details.getFile()
-            assert expectedVisitations[0].contains(file)
-            expectedVisitations[0].remove(file)
-            if (expectedVisitations[0].isEmpty() || file == stopOn) {
-                expectedVisitations.remove(0)
-                if (file == stopOn) {
-                    details.stopVisiting()
-                }
+            def found = expectedVisitations.find { it.remove(file) }
```
Comment (wolfs) [inline subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy]: Let's not modify something within the find.

Comment (wolfs) [inline subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy]: We need to assert that we found something.

File: subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy [hunk 21]
```diff
@@ -234,23 +263,26 @@ class DirectoryFileTreeTest extends AbstractProjectBuilderSpec {
...
+            if (file == stopOn) {
+                expectedVisitations.remove(found)
+                details.stopVisiting()
+            } else if (found.isEmpty()) {
+                expectedVisitations.remove(found)
```
Comment (wolfs) [inline subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy]: ❌ Does it make sense any more to provide a list of lists of expected visitations here if we don't expect the order anyway? We could directly use a collecting file visitor and then assert that visited files at the end.

Comment (paplorinc) [reply subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy]: I didn't make sense in the beginning either, just didn't want to refactor it too much. Done

File: subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy (hunk 20 above)
Comment (paplorinc) [reply subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy]: removed

Comment (paplorinc) [reply subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy]: removed

File: subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy (hunk 19 above)
Comment (paplorinc) [reply subprojects/files/src/test/groovy/org/gradle/api/internal/file/collections/DirectoryFileTreeTest.groovy]: done

**Comment by paplorinc** [comment]
Merged, thank you for the valuable reviews.
If there's anything that still needs to be added, let me know and I'll do it in a separate PR.
